pkg/controller/storagecluster/kms_resources.go

```diff
@@ -0,0 +1,86 @@
+package storagecluster
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	ocsv1 "github.com/openshift/ocs-operator/pkg/apis/ocs/v1"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	// KMSConfigMapName is the name configmap which has KMS config details
+	KMSConfigMapName = "ocs-kms-connection-details"
+	// KMSTokenSecretName is the name of the secret which has KMS token details
+	KMSTokenSecretName = "ocs-kms-token"
+	// KMSProviderKey is the key in config map to get the KMS provider name
+	KMSProviderKey = "KMS_PROVIDER"
+	// VaultKMSProvider a constant to represent 'vault' KMS provider
+	VaultKMSProvider = "vault"
+)
+
+var (
+	// currently supported KMS providers mapped to their address key
+	kmsProviderAddressKeyMap = map[string]string{
+		VaultKMSProvider: "VAULT_ADDR",
+	}
+)
+
+// kmsConfigMapValidateFunc is a functional type,
+// which validates the given configmap and returns an error if any
+type kmsConfigMapValidateFunc func(*corev1.ConfigMap) error
+
+var (
+	// just to make sure 'isKMSProviderReachable' function follows the validate function type
+	_ kmsConfigMapValidateFunc = isKMSProviderReachable
+)
+
+// getKMSConfigMap function try to return a KMS ConfigMap.
+// if 'kmsValidateFunc' function is present it try to validate the retrieved config map.
+func getKMSConfigMap(instance *ocsv1.StorageCluster, client client.Client, kmsValidateFunc kmsConfigMapValidateFunc) (*corev1.ConfigMap, error) {
+	// if 'KMS' is not enabled, nothing to fetch
+	if !instance.Spec.Encryption.KeyManagementService.Enable {
+		return nil, nil
+	}
+	kmsConfigMap := corev1.ConfigMap{}
+	err := client.Get(context.TODO(),
+		types.NamespacedName{
+			Name:      KMSConfigMapName,
+			Namespace: instance.ObjectMeta.Namespace,
+		},
+		&kmsConfigMap,
+	)
+	// if validation function is provided, use it
+	if kmsValidateFunc != nil {
+		err = kmsValidateFunc(&kmsConfigMap)
+	}
+	return &kmsConfigMap, err
+}
+
+// isKMSProviderReachable function checks whether the provided address is reachable or not.
+// This function won't validate any other cases and only returns an error if the provided
+// KMS provider address is not reachable. All other validations will be done on rook side.
+func isKMSProviderReachable(kmsConfigMap *corev1.ConfigMap) error {
```

Comment:
A better func signature would be to return a bool as well as the err

Reply:
+1

Reply:
@nbalacha , the bool value returned won't have much use, right?. Return `error` itself will show that whether reachable function is successful (absence of error) or not (presence of error), and this is a common practice in golang.

In the case where no error is being raised during the execution of function (something like `isNameSame(name1, name2 string) bool`), yes, then definitely return a bool value (no need of an extra error).

A scenario where I would have used multiple return values would be, if I was returning a set of URLs which are not reachable along with the error.

A more strict go convention would be to remove `is` prefix from `isKMSProviderReachable` and make the function name further simpler (like `kmsProviderReachable`). Then we use multiple `is` prefixed functions, something like `isValidURL(err error) bool` or `isValidProtocol(err error) bool` etc; to figure out what is the reason behind the returned error.

Golang's [os package](https://golang.org/pkg/os/) has more such examples.

Reply:
> A more strict go convention would be to remove `is` prefix from `isKMSProviderReachable` and make the function name further simpler (like `kmsProviderReachable`). Then we use multiple `is` prefixed functions, something like `isValidURL(err error) bool` or `isValidProtocol(err error) bool` etc; to figure out what is the reason behind the returned error.

That also sounds good!

Reply:
I would still prefer adding a bool return for cleaner function definition - an Is<do-something> implies a bool value. 
I checked the os package and the only functions I could see which return only an error are those that perform actions - chmod, chdir etc etc. That would hold in this case if the function was called checkKMSProvider or similar.

Reply:
@nbalacha , your point is fair, as the function name starts with an `is***` we expect it to have a boolean return value.
Changing the function name to `func reachKMSProvider(configmap) error`. Hope this is fine.